File: restdoc-client/src/main/java/restdoc/client/RestDocClientConfiguration.java
package restdoc.client;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.env.Environment;
import org.springframework.web.client.RestTemplate;
import restdoc.client.executor.HttpTaskExecutor;
import restdoc.client.remoting.ApplicationClient;
import restdoc.client.remoting.HttpTaskRequestProcessor;
import restdoc.client.remoting.PostEmptyApiTemplateRequestProcessor;

@Configuration
@EnableConfigurationProperties(RestDocProperties.class)
public class RestDocClientConfiguration {

    private static Logger log = LoggerFactory.getLogger(RestDocClientConfiguration.class);

    @Bean
    @ConditionalOnMissingBean
    public ApplicationClient applicationClient(RestDocProperties restDocProperties,
                                               HttpTaskRequestProcessor httpTaskRequestProcessor,
                                               PostEmptyApiTemplateRequestProcessor postEmptyApiTemplateRequestProcessor
                                               ) {
        ApplicationClient applicationClient = new ApplicationClient(
                restDocProperties,
                httpTaskRequestProcessor,
                postEmptyApiTemplateRequestProcessor);

        applicationClient.connection();
        return applicationClient;
    }

    @Bean
    @ConditionalOnMissingBean
    public HttpTaskExecutor httpTaskExecutor(RestTemplate restTemplate, Environment environment) {
        return new HttpTaskExecutor(restTemplate, environment);
    }

    @Bean
    @ConditionalOnMissingBean
    public HttpTaskRequestProcessor httpTaskRequestProcessor(HttpTaskExecutor httpTaskExecutor) {
        return new HttpTaskRequestProcessor(httpTaskExecutor);
    }

    @Bean
    @ConditionalOnMissingBean
    public RestTemplate restTemplate() {
        return new RestTemplate();
    }

    @Bean
    @ConditionalOnMissingBean
    public EndpointsListener endpointsListener() {
        return new EndpointsListener();
    }

    @Bean
    @ConditionalOnMissingBean
    public PostEmptyApiTemplateRequestProcessor postEmptyApiTemplateRequestProcessor() {
        return new PostEmptyApiTemplateRequestProcessor();
    }
}

File: restdoc-rpc-client/restdoc-restweb-client/src/main/java/restdoc/client/restweb/config/RestDocClientConfiguration.kt
package restdoc.client.restweb.config

import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.boot.context.properties.EnableConfigurationProperties
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.env.Environment
import org.springframework.web.client.RestTemplate
import restdoc.client.restweb.context.EndpointsListener
import restdoc.client.restweb.invoke.HttpInvoker
import restdoc.client.restweb.remoting.HttpApplicationAgent
import restdoc.client.restweb.remoting.HttpTaskRequestProcessor
import restdoc.client.restweb.remoting.PostEmptyApiTemplateRequestProcessor

@Configuration
@EnableConfigurationProperties(RestDocProperties::class)
open class RestDocClientConfiguration {

    @Bean @ConditionalOnMissingBean
    open fun applicationClient(restDocProperties: RestDocProperties,
                               httpTaskRequestProcessor: HttpTaskRequestProcessor,
                               postEmptyApiTemplateRequestProcessor: PostEmptyApiTemplateRequestProcessor
    ): HttpApplicationAgent {
        val httpApplicationAgent = HttpApplicationAgent(
                restDocProperties,
                httpTaskRequestProcessor,
                postEmptyApiTemplateRequestProcessor)
        httpApplicationAgent.connection()
        return httpApplicationAgent
    }

    @Bean @ConditionalOnMissingBean
    open fun httpTaskExecutor(restTemplate: RestTemplate, environment: Environment) = HttpInvoker(restTemplate, environment)

    @Bean @ConditionalOnMissingBean
    open fun httpTaskRequestProcessor(httpInvoker: HttpInvoker?) = HttpTaskRequestProcessor(httpInvoker)

    @Bean @ConditionalOnMissingBean
    open fun restTemplate() = RestTemplate()

    @Bean @ConditionalOnMissingBean
    open fun endpointsListener(environment: Environment?) = EndpointsListener(environment)

    @Bean @ConditionalOnMissingBean
    open fun postEmptyApiTemplateRequestProcessor(endpointsListener: EndpointsListener) = PostEmptyApiTemplateRequestProcessor(endpointsListener)

}